lib/smack/NormalizeLoops.cpp

```diff
@@ -0,0 +1,134 @@
+//
+// This file is distributed under the MIT License. See LICENSE for details.
+//
+
+//
+// This pass normalizes loop structures to allow easier generation of Boogie
+// code when a loop edge comes from a branch. Specifically, a forwarding block
+// is added between a branching block and the loop header.
+//
+
+#include "smack/NormalizeLoops.h"
+#include "smack/Naming.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/ValueSymbolTable.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/Analysis/LoopInfo.h"
+#include "llvm/IR/Dominators.h"
+#include <map>
+#include <set>
+
+namespace smack {
+
+using namespace llvm;
+
+// Register LoopInfo
+void NormalizeLoops::getAnalysisUsage(AnalysisUsage& AU) const {
+  AU.addRequired<LoopInfoWrapperPass>();
+}
+
+/**
+ * Creates a basic block in \param func which unconditionally branches to
+ * \param target.
+ */
+BasicBlock* makeForwardingBlock(BasicBlock* target, Function* func) {
+  auto& context = target->getContext();
+  BasicBlock* result = BasicBlock::Create(context, "forwarder", func);
+  BranchInst* branch = BranchInst::Create(target);
+  result->getInstList().push_back(branch);
+  return result;
+}
+
+/**
+ * Splits any edge between \param BB and \param headerBlock in a loop body,
+ * by adding a forwarding block between \param BB and \param headerBlock. The
+ * \param blockMap is updated to map between \param BB and the newly created
+ * forwarding block.
+ */
+void splitBlock(BasicBlock* BB, BasicBlock* headerBlock, Function* func,
+                std::map<BasicBlock*, BasicBlock*>& blockMap) {
+  TerminatorInst* ti = BB->getTerminator();
+  unsigned numSuccs = ti->getNumSuccessors();
+  if (numSuccs <= 1) { return; }
+  BasicBlock* forwarder = nullptr;
+  for (unsigned i = 0; i < numSuccs; ++i) {
+    BasicBlock* succ = ti->getSuccessor(i);
+    if (succ != headerBlock) { continue; }
+
+    // Successor is the header block and this block has multiple successors,
+    // so it needs a forwarding block.
+    if (forwarder == nullptr) {
```

Comment:
Declaring `forwarder` outside of the loop, and checking this condition here seems fairly convoluted, given that this control location is only reachable for one single block: `headerBlock`. This condition should thus be guaranteed to be true, and there should be no reason to initialize `forwarder` to `nullptr`. In fact, the only reason for this loop is to determine the successor index of `headerBlock`, if any. This code would be much clearer with a call to some `getSuccessorIndex(ti, headerBlock)` function, which returns the successor index of `headerBlock`, or `-1` if its not a successor of `ti`.

Reply:
This is to deal with switch instructions, which may point to the header more than once.

Reply:
Oh ok, fair point. But this subtlety that you point out can not be inferred by reading this code, unless you already have switch instructions in mind. It would thus help future maintenance to make this clearer. I would suggest a variation on my previous suggestion: a call to some `getSuccessorIndices(ti, headerBlock)` function, which returns the successor indices of `headerBlock`. You can then return early if this list is empty, replacing the check of `numSuccs <= 1`, and otherwise create the forwarding block.

Reply:
This is done. However, it still may be desirable to check the number of successors to prevent a forwarding block from being created between a body block that branches unconditionally to the loop header.